README.rst
Flask-URS
==============

.. image:: https://travis-ci.org/justinwp/flask-urs.svg?branch=master
    :target: https://travis-ci.org/justinwp/flask-urs

.. image:: https://img.shields.io/pypi/v/flask-urs.svg
    :target: https://pypi.python.org/pypi/Flask-URS/
    :alt: Latest Version

.. image:: https://coveralls.io/repos/github/justinwp/flask-urs/badge.svg?branch=master
    :target: https://coveralls.io/github/justinwp/flask-urs?branch=master

.. image:: https://img.shields.io/pypi/dm/flask-urs.svg
    :target: https://pypi.python.org/pypi//Flask-URS/
    :alt: Downloads

.. image:: https://img.shields.io/pypi/l/flask-urs.svg
    :target: https://pypi.python.org/pypi/Flask-URS/
    :alt: License

`NASA EarthData(URS) <https://urs.earthdata.nasa.gov/documentation>`_ Oauth2 Login with JWT (JSON Web Tokens) for Flask applications

Heavily influenced by `Flask-JWT <https://github.com/mattupstate/flask-jwt>`_

Resources
---------

- `Issue Tracker <http://github.com/justinwp/flask-urs/issues>`_
- `Code <http://github.com/justinwp/flask-urs/>`_


CHANGES
Flask-URS Changelog
========================

Here you can see the full list of changes between each Flask-URS release.

Version 0.1.2
-------------

Released March 11th 2016

- Copy/Replace Fixes

Version 0.1.0
-------------

Released March 9th 2016

- Initial release


tests/conftest.py
# -*- coding: utf-8 -*-
"""
    tests.conftest
    ~~~~~~~~~~~~~~

    Test fixtures and what not
"""

from datetime import timedelta

import pytest

from flask import Flask, jsonify

import flask_urs

import responses


@pytest.fixture(scope='function')
def urs():
    return flask_urs.URS()


@pytest.fixture(scope='function')
def user():
    return {
        "email_address": "test@email.com",
        "uid": "username",
        "affiliation": "Government",
        "organization": "Somewhere",
        "first_name": "First",
        "last_name": "Last",
        "user_type": "Science Team",
        "country": "United States"
    }


@pytest.fixture(scope='function')
def app(urs, user):
    app = Flask(__name__)
    app.debug = True
    app.config['SECRET_KEY'] = 'super-secret'
    app.config['JWT_EXPIRATION_DELTA'] = timedelta(milliseconds=600)
    app.config['JWT_EXPIRATION_LEEWAY'] = timedelta(milliseconds=5)

    urs.init_app(app)

    @app.route('/protected')
    @flask_urs.jwt_required()
    def protected():
        return 'success'

    @app.route("/authorize")
    def authorize():
        return jsonify({
            "access_token": "asdf",
            "refresh_token": "asdf",
            "endpoint": "/endpoint"
        })

    @app.route("/endpoint")
    def endpoint():
        return jsonify(user)

    # Establish an application context before running the tests.
    ctx = app.app_context()
    ctx.push()

    def teardown():
        ctx.pop()

    return app


@pytest.fixture(scope='function')
def client(app):
    return app.test_client()


@pytest.fixture(scope='function')
def fake_oauth_success(urs, app, user):
    token_response = {"access_token": "asdf", "endpoint": "api/username"}

    responses.add(responses.POST, urs._token_url, json=token_response, status=200)

    responses.add(responses.GET, app.config.get("URS_HOST") + token_response["endpoint"],
                  json=user, status=200)


tests/test_jwt.py
# -*- coding: utf-8 -*-
"""
    tests.test_jwt
    ~~~~~~~~~~~~~~

    Flask-URS-JWT tests
"""
import time

from itsdangerous import TimedJSONWebSignatureSerializer

from flask import Flask, json, jsonify

import responses

import pytest

import flask_urs


def post_json(client, url, data):
    resp = client.post(
        url,
        headers={'content-type': 'application/json'},
        data=json.dumps(data)
    )
    return resp, json.loads(resp.data)


def assert_error_response(r, code, msg, desc):
    assert r.data is not None
    try:
        jdata = json.loads(r.data)
    except:
        print(r.data)
        return
    assert r.status_code == code
    assert jdata['status_code'] == code
    assert jdata['error'] == msg
    assert jdata['description'] == desc


def test_initialize():
    app = Flask(__name__)
    app.config['SECRET_KEY'] = 'super-secret'
    urs = flask_urs.URS(app)
    assert isinstance(urs, flask_urs.URS)
    assert len(app.url_map._rules) == 2


def test_jwt_required_decorator_with_valid_token(urs, client, user):
    token = urs.encode_callback(user)
    resp = client.get(
        '/protected',
        headers={'authorization': 'Bearer ' + token})
    assert resp.status_code == 200
    assert resp.data == b'success'


def test_jwt_required_decorator_with_valid_request_current_user(urs, client, user):
    with client as c:
        token = urs.encode_callback(user)
        c.get(
            '/protected',
            headers={'authorization': 'Bearer ' + token})
        assert flask_urs.current_user


def test_jwt_required_decorator_with_invalid_request_current_user(app, client):
    with client as c:
        c.get(
            '/protected',
            headers={'authorization': 'Bearer bogus'})
        assert not flask_urs.current_user


def test_jwt_required_decorator_with_invalid_authorization_headers(client):
    # Missing authorization header
    r = client.get('/protected')
    assert_error_response(r, 401, 'Authorization Required', 'Authorization header was missing')
    assert r.headers['WWW-Authenticate'] == 'JWT realm="Login Required"'

    # Not a bearer token
    r = client.get('/protected', headers={'authorization': 'Bogus xxx'})
    assert_error_response(r, 400, 'Invalid JWT header', 'Unsupported authorization type')

    # Missing token
    r = client.get('/protected', headers={'authorization': 'Bearer'})
    assert_error_response(r, 400, 'Invalid JWT header', 'Token missing')

    # Token with spaces
    r = client.get('/protected', headers={'authorization': 'Bearer xxx xxx'})
    assert_error_response(r, 400, 'Invalid JWT header', 'Token contains spaces')


def test_jwt_required_decorator_with_invalid_jwt_tokens(client, user, urs):
    token = urs.encode_callback(user)

    # Undecipherable
    r = client.get('/protected', headers={'authorization': 'Bearer %sX' % token})
    assert_error_response(r, 400, 'Invalid JWT', 'Token is undecipherable')

    # Expired
    time.sleep(1)
    r = client.get('/protected', headers={'authorization': 'Bearer ' + token})
    assert_error_response(r, 400, 'Invalid JWT', 'Token is expired')


def test_jwt_required_decorator_with_missing_user(urs, client, user):
    token = urs.encode_callback(user)

    @urs.user_handler
    def load_user(payload):
        return None

    r = client.get('/protected', headers={'authorization': 'Bearer %s' % token})
    assert_error_response(r, 400, 'Invalid JWT', 'User does not exist')


@responses.activate
@pytest.mark.usefixtures("fake_oauth_success")
def test_custom_response_handler(app, client, urs, user):
    @urs.response_handler
    def response_callback(user, jwt, access):
        return jsonify({"jwt": jwt})

    r = client.get(
        app.config.get("URS_URL_PREFIX") + app.config.get("URS_CALLBACK_RULE") + "?code=x")

    assert r.status_code == 200
    assert "jwt" in json.loads(r.data)


def test_default_encode_handler(user, app, urs):
    token = urs.encode_callback(user)

    serializer = TimedJSONWebSignatureSerializer(
        secret_key=app.config['JWT_SECRET_KEY']
    )
    decoded = serializer.loads(token)
    assert decoded['email_address'] == user['email_address']


def test_custom_encode_handler(urs, user, app):
    serializer = TimedJSONWebSignatureSerializer(
        app.config['JWT_SECRET_KEY'],
        algorithm_name=app.config['JWT_ALGORITHM']
    )

    @urs.encode_handler
    def encode_data(payload):
        return serializer.dumps({'foo': 42}).decode('utf-8')

    token = urs.encode_callback(user)

    decoded = serializer.loads(token)
    assert decoded == {'foo': 42}


def test_custom_decode_handler(client, user, urs):
    @urs.decode_handler
    def decode_data(data):
        return {'user_id': user['uid']}

    with client as c:
        token = urs.encode_callback(user)

        c.get(
            '/protected',
            headers={'authorization': 'Bearer ' + token})
        assert flask_urs.current_user == decode_data(user)


def test_custom_payload_handler(client, urs, user):
    @urs.user_handler
    def load_user(payload):
        if payload['uid'] == user['uid']:
            return user

    @urs.payload_handler
    def make_payload(u):
        return {
            'id': u['uid']
        }

    with client as c:
        token = urs.encode_callback(user)

        c.get(
            '/protected',
            headers={'authorization': 'Bearer ' + token})
        assert flask_urs.current_user == user


example/app.py
from flask import Flask, render_template, request, jsonify
from flask_urs import URS, URSError
import os

app = Flask(__name__)
app.debug = True
app.config['SECRET_KEY'] = 'super-secret'
app.config['URS_CLIENT_ID'] = os.environ['URS_CLIENT_ID']
app.config['URS_UID'] = os.environ['URS_UID']
app.config['URS_PASSWORD'] = os.environ['URS_PASSWORD']

urs = URS(app)


@app.route('/')
def protected():
    redirect_uri = request.base_url[:-1] + urs.redirect_url_rule
    return render_template('urs.html', client_id=app.config.get('URS_CLIENT_ID'),
                           redirect_uri=redirect_uri)


@urs.response_handler
def response_callback(user, jwt, access):
    return render_template('custom_callback.html', jwt=jwt)


@urs.payload_handler
def payload_callback(user):
    user.update({'custom_data': 'asdfasdfasdfsdfsd'})
    return user


@app.errorhandler(URSError)
def handle_invalid_usage(error):
    response = jsonify(error.__dict__)
    response.status_code = error.status_code
    return response


if __name__ == '__main__':
    app.run()


flask_urs/__init__.py
# -*- coding: utf-8 -*-
"""
    flask_urs_urs
    ~~~~~~~~~

    Flask-URS-JWT module
"""

from flask import current_app, render_template, Blueprint, request, jsonify

from flask import _app_ctx_stack as stack
from itsdangerous import (
    TimedJSONWebSignatureSerializer,
    SignatureExpired,
    BadSignature
)
from functools import wraps
from werkzeug.local import LocalProxy
import requests
from datetime import timedelta
from collections import OrderedDict

__version__ = '0.1.3'

current_user = LocalProxy(lambda: getattr(stack.top, 'current_user', None))

_urs = LocalProxy(lambda: current_app.extensions['urs'])

CONFIG_DEFAULTS = {
    'URS_CALLBACK_RULE': '/callback',
    'URS_URL_PREFIX': '/urs',
    'URS_CALLBACK_TEMPLATE': 'urs/callback.html',
    'JWT_EXPIRATION_DELTA': 3600,
    'JWT_EXPIRATION_LEEWAY': 100,
    'JWT_VERIFY_EXPIRATION': True,
    'JWT_ALGORITHM': 'HS256',
    'JWT_DEFAULT_REALM': 'Login Required',
    'URS_HOST': 'https://urs.earthdata.nasa.gov/',
    'URS_TOKEN_PATH': 'oauth/token'
}


def _get_serializer():
    expires_in = current_app.config['JWT_EXPIRATION_DELTA']
    leeway = current_app.config['JWT_EXPIRATION_LEEWAY']
    if isinstance(expires_in, timedelta):
        expires_in = int(expires_in.total_seconds())
    if isinstance(leeway, timedelta):
        leeway = int(leeway.total_seconds())
    expires_in_total = expires_in + leeway
    return TimedJSONWebSignatureSerializer(
        secret_key=current_app.config['JWT_SECRET_KEY'],
        expires_in=expires_in_total,
        algorithm_name=current_app.config['JWT_ALGORITHM']
    )


def jwt_required(realm=None):
    """View decorator that requires a valid JWT token to be present in the request

    :param realm: an optional realm
    """

    def wrapper(fn):
        @wraps(fn)
        def decorator(*args, **kwargs):
            verify_jwt(realm)
            return fn(*args, **kwargs)

        return decorator

    return wrapper


class JWTError(Exception):
    def __init__(self, error, description, status_code=400, headers=None):
        self.error = error
        self.description = description
        self.status_code = status_code
        self.headers = headers


class URSError(Exception):
    def __init__(self, error, description, status_code=400, headers=None):
        self.error = error
        self.description = description
        self.status_code = status_code
        self.headers = headers


def verify_jwt(realm=None):
    """Does the actual work of verifying the JWT data in the current request.
    This is done automatically for you by `jwt_required()` but you could call it manually.
    Doing so would be useful in the context of optional JWT access in your APIs.

    :param realm: an optional realm
    """
    realm = realm or current_app.config['JWT_DEFAULT_REALM']
    auth = request.headers.get('Authorization', None)

    if auth is None:
        raise JWTError('Authorization Required', 'Authorization header was missing', 401, {
            'WWW-Authenticate': 'JWT realm="%s"' % realm
        })

    parts = auth.split()

    if parts[0].lower() != 'bearer':
        raise JWTError('Invalid JWT header', 'Unsupported authorization type')
    elif len(parts) == 1:
        raise JWTError('Invalid JWT header', 'Token missing')
    elif len(parts) > 2:
        raise JWTError('Invalid JWT header', 'Token contains spaces')

    try:
        handler = _urs.decode_callback
        payload = handler(parts[1])
    except SignatureExpired:
        raise JWTError('Invalid JWT', 'Token is expired')
    except BadSignature:
        raise JWTError('Invalid JWT', 'Token is undecipherable')

    stack.top.current_user = user = _urs.user_callback(payload)

    if user is None:
        raise JWTError('Invalid JWT', 'User does not exist')


def _default_payload_handler(user):
    stack.top.current_user = user
    return user


def _default_user_handler(payload):
    return payload


def _default_encode_handler(payload):
    """Return the encoded payload."""
    return _get_serializer().dumps(payload).decode('utf-8')


def _default_decode_handler(token):
    """Return the decoded token."""
    try:
        result = _get_serializer().loads(token)
    except SignatureExpired:
        if current_app.config['JWT_VERIFY_EXPIRATION']:
            raise
    return result


def _default_jwt_error_handler(error):
    return jsonify(OrderedDict([
        ('status_code', error.status_code),
        ('error', error.error),
        ('description', error.description),
    ])), error.status_code, error.headers


def _default_response_handler(user, jwt, access):
    return render_template(CONFIG_DEFAULTS.get('URS_CALLBACK_TEMPLATE'), jwt=jwt)


class URS(object):
    def __init__(self, app=None):
        self.user_callback = _default_user_handler
        self.response_callback = _default_response_handler
        self.encode_callback = _default_encode_handler
        self.decode_callback = _default_decode_handler
        self.payload_callback = _default_payload_handler
        self.jwt_error_callback = _default_jwt_error_handler

        if app is not None:
            self.init_app(app)

    def init_app(self, app):
        for k, v in CONFIG_DEFAULTS.items():
            app.config.setdefault(k, v)

        app.config.setdefault('JWT_SECRET_KEY', app.config['SECRET_KEY'])

        bp = Blueprint('urs_urs', __name__, template_folder='templates')
        bp.url_prefix = app.config.get('URS_URL_PREFIX', '')
        bp.add_url_rule(app.config.get('URS_CALLBACK_RULE'), methods=['GET'],
                        view_func=self.callback)

        app.register_blueprint(bp)

        if not hasattr(app, 'extensions'):  # pragma: no cover
            app.extensions = {}

        app.errorhandler(JWTError)(self.jwt_error_callback)

        app.extensions['urs'] = self

    @property
    def redirect_url_rule(self):
        return current_app.config.get('URS_URL_PREFIX') + current_app.config.get(
            'URS_CALLBACK_RULE')

    @property
    def _token_url(self):
        return current_app.config.get('URS_HOST', CONFIG_DEFAULTS['URS_HOST']) \
               + current_app.config.get('URS_TOKEN_PATH', CONFIG_DEFAULTS['URS_TOKEN_PATH'])

    def callback(self):
        code = request.args.get('code', None)
        if code is None:
            raise URSError('Invalid Code', 'No Authorization Code in Arguments')

        access = self.get_token(code)
        user = self.get_user(access['access_token'], access['endpoint'])
        payload = self.payload_callback(user)
        jwt = self.encode_callback(payload)

        return self.response_callback(user, jwt, access)

    def refresh(self, refresh_token):
        """
        Get new access token...
        :param refresh_token:
        :return:
        """

        # TODO

        pass

    def get_user(self, token, endpoint, refresh_token=None):
        headers = {
            "Authorization": "Bearer %s" % token
        }

        r = requests.get(current_app.config.get('URS_HOST') + endpoint, headers=headers)

        if r.status_code != 200:
            raise URSError('Invalid Code', 'No Authorization Code')

        return r.json()

    def get_token(self, code):
        headers = {
            "Content-Type": "application/x-www-form-urlencoded"
        }

        data = {
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": request.base_url
        }

        auth = requests.auth.HTTPBasicAuth(current_app.config.get('URS_UID'),
                                           current_app.config.get('URS_PASSWORD'))

        r = requests.post(self._token_url, headers=headers, data=data, auth=auth)

        if r.status_code == 401:
            raise URSError('Token Access Denied', 'Incorrect Application UID or Password',
                           status_code=500)

        elif r.status_code == 400:
            error = r.json()
            raise URSError(r.json()['error'], error['error_description'], status_code=500)

        elif r.status_code != 200:
            raise URSError('Unknown Error', 'Could Not Retrieve Access Token', status_code=500)

        return r.json()

    def response_handler(self, callback):
        """Specifies the response handler function. This function receives a
        JWT-encoded payload, the token payload and returns a Flask response.

        :param callable callback: the response handler function
        """
        self.response_callback = callback
        return callback

    def user_handler(self, callback):
        """Specifies the user handler function. This function receives the token payload as
        its only positional argument. It should return an object representing the current
        user. Example::

            @urs.user_handler
            def load_user(payload):
                if payload['user_id'] == 1:
                    return User(id=1, username='joe')

        :param callback: the user handler function
        """
        self.user_callback = callback
        return callback

    def error_handler(self, callback):
        """Specifies the error handler function. This function receives a URSError instance as
        its only positional argument. It can optionally return a response. Example::

            @urs.error_handler
            def error_handler(e):
                return "Something bad happened", 400

        :param callback: the error handler function
        """
        self.error_callback = callback
        return callback

    def encode_handler(self, callback):
        """Specifies the encoding handler function. This function receives a
        payload and signs it.

        :param callable callback: the encoding handler function
        """
        self.encode_callback = callback
        return callback

    def decode_handler(self, callback):
        """Specifies the decoding handler function. This function receives a
        signed payload and decodes it.

        :param callable callback: the decoding handler function
        """
        self.decode_callback = callback
        return callback

    def payload_handler(self, callback):
        """Specifies the payload handler function. This function receives a
        user object and returns a dictionary payload.

        Example::

            @urs.payload_handler
            def make_payload(user):
                return {
                    'user_id': user.id,
                    'exp': datetime.utcnow() + current_app.config['JWT_EXPIRATION_DELTA']
                }

        :param callable callback: the payload handler function
        """
        self.payload_callback = callback
        return callback

    def jwt_error_handler(self, callback):
        """Specifies the error handler function. Example::
            @urs.error_handler
            def error_handler(e):
                return "Something bad happened", 400
        :param callback: the error handler function
        """
        self.jwt_error_callback = callback
        return callback
